extensions/poetry/main.ts

```diff
@@ -0,0 +1,108 @@
+import { PhylumApi } from "phylum"
+import { parseDryRun } from "./parse.ts"
+import { red, green, yellow } from "https://deno.land/std@0.150.0/fmt/colors.ts";
+
+class FileBackup {
+  readonly fileName: string
+  readonly fileContent: string | null
+
+  constructor(fileName: string) {
+    this.fileName = fileName
+    this.fileContent = null
+  }
+
+  async backup() {
+    try {
+      this.fileContent = await Deno.readTextFile(this.fileName)
+    } catch (e) { }
+  }
+
+  async restoreOrDelete() {
+    try {
+      if (this.fileContent != null) {
+        await Deno.writeTextFile(this.fileName, this.fileContent)
+      } else {
+        await Deno.remove(this.fileName)
+      }
+    } catch (e) { }
+  }
+}
+
+// Parse the output of `poetry` subcommands that support the `--dry-run` flag,
+// and submit the packages.
+//
+// Add the `-vvv` flags to get a detailed report of the dependency resolution
+// process. This way, we can track the actual lockfile changes, as specifying
+// `--dry-run` only would not output anything in combination with flags such
+// as `--lock` that do not perform the actual operations.
+async function poetryCheckDryRun(subcommand: string, args: string[]): boolean {
+  // Read and backup the current poetry lockfile contents.
+  const lockfileBackup = new FileBackup('poetry.lock')
+  const manifestBackup = new FileBackup('pyproject.toml')
+
+  await lockfileBackup.backup()
+  await manifestBackup.backup()
+
+  let process = Deno.run({
+    cmd: ['poetry', subcommand, '-vvv', '-n', '--dry-run', ...args.map(s => s.toString())],
+    stdout: 'piped',
+    stderr: 'piped',
+  })
```

Comment:
The whole point is that this shouldn't update the lockfile, right? Why do you store and revert it?
If you need to store and revert it anyway, what's the point of doing a dry run?

There's just so much complexity in such a simple extension, it seems somewhat difficult to get a reasonable grasp of all the invariants and how they play together.

Reply:
I mentioned this elsewhere: `poetry` actually applies changes to the lockfile and `pyproject.toml` _regardless_ of the `--dry-run` and `--lock` flags. It is a [known bug](https://github.com/python-poetry/poetry/issues/3666) that should be addressed in 1.2, which has been delayed for quite a while and is not going to be released anytime soon afaik. Forcing backup and restore is the only workaround to that.

I agree that the extension is more complex than I'd like but the workarounds are definitely necessary.